src/sinks/mqtt/sink.rs

```diff
@@ -0,0 +1,205 @@
+    #[snafu(display("TLS error: {}", source))]
+    Tls { source: TlsError },
+    #[snafu(display("MQTT client error: {}", source))]
+    Client { source: ClientError },
+}
+
+#[derive(Clone)]
+pub struct MqttConnector {
+    options: MqttOptions,
+    topic: Template,
+}
+
+impl MqttConnector {
+    pub fn new(options: MqttOptions, topic: String) -> Result<Self, MqttError> {
+        let topic = Template::try_from(topic).context(TopicTemplateSnafu)?;
+        Ok(Self { options, topic })
+    }
+
+    fn connect(&self) -> (AsyncClient, EventLoop) {
+        AsyncClient::new(self.options.clone(), 1024)
+    }
+
+    pub async fn healthcheck(&self) -> crate::Result<()> {
+        // FIXME: Right now there is no way to implement the healthcheck properly: https://github.com/bytebeamio/rumqtt/issues/562
+        Ok(())
+    }
+}
+
+pub struct MqttSink {
+    transformer: Transformer,
+    encoder: Encoder<()>,
+    connector: MqttConnector,
+    finalizers_queue: VecDeque<EventFinalizers>,
+    quality_of_service: MqttQoS,
+    events_sent: Registered<EventsSent>,
+    bytes_sent: Registered<BytesSent>,
+}
+
+impl MqttSink {
+    pub fn new(config: &MqttSinkConfig, connector: MqttConnector) -> crate::Result<Self> {
+        let transformer = config.encoding.transformer();
+        let serializer = config.encoding.build()?;
+        let encoder = Encoder::<()>::new(serializer);
+        let finalizers_queue = VecDeque::new();
+
+        Ok(Self {
+            transformer,
+            encoder,
+            connector,
+            finalizers_queue,
+            quality_of_service: config.quality_of_service,
+            events_sent: register!(EventsSent::from(Output(None))),
+            bytes_sent: register!(BytesSent::from(Protocol("mqtt".into()))),
+        })
+    }
+
+    /// outgoing events main loop
+    async fn handle_events<I>(
+        &mut self,
+        input: &mut I,
+        client: &mut AsyncClient,
+        connection: &mut EventLoop,
+    ) -> Result<(), ()>
+    where
+        I: Stream<Item = Event> + Unpin,
+    {
+        loop {
+            tokio::select! {
+                // handle connection errors
+                msg = connection.poll() => {
+                    match msg {
+                        Ok(rumqttc::Event::Outgoing(rumqttc::Outgoing::PubRel(_))) => {
+                            // publish has been acknowledged by the MQTT server
+                            if let Some(finalizers) = self.finalizers_queue.pop_front() {
+                                finalizers.update_status(EventStatus::Delivered);
+                            }
+                        }
+                        Ok(_) => {}
+                        Err(error) => {
+                            emit!(MqttConnectionError { error });
+                            return Err(());
+                        }
+                    }
+                },
+
+                // handle outgoing events
+                event = input.next() => {
```

Comment:
I think it would help readability a lot to extract the body of the msg , and event cases into helper functions.

Reply:
I would be honest - I have tried to do it, but fmpov I got an even worse result after the refactoring attempt. I have tried to find a good `tokio::select!` pattern in Vector but after a quick look through I see that in almost all cases is used the same pattern - describe multiple futures inside `tokio::select!` macro.

I am not against this refactoring, ofc. I just cannot do better here.